doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+    $ cd $INSTALL_LOCATION
+    $ git clone https://github.com/$GITHUB_USERNAME/mne-python.git
+
+Finally, set up a link between your local clone and the official repository
+("upstream"):
+
+.. code-block:: console
+
+    $ cd mne-python
+    $ git remote add upstream git://github.com/mne-tools/mne-python.git
+    $ git fetch --all
+
+.. note::
+
+    We use ``git://`` instead of ``https://`` in the address for the official
+    "upstream" remote repository. ``git://`` addresses are read-only, which
+    means you can *pull* the official repository into your local copy (in order
+    to stay up-to-date with changes made by other contributors) but you cannot
+    *push* anything from your computer directly into the upstream remote.
+    Instead, you must push your changes to your own remote fork first, and then
+    create a pull request from your remote into the upstream remote (and even
+    then, your changes are not automatically accepted into the upstream; the
+    changes must be *approved* by maintainers and then *merged* into upstream).
+
+Next, use the `environment file`_ provided in the root of the MNE-Python
+repository to set up your local development environment. This will install all
+of the dependencies needed for running MNE-Python. The following commands
+create the conda environment:
+
+.. code-block:: console
+
+    $ PREFERRED_ENVIRONMENT_NAME="mnedev"
+    $ conda env create -n $PREFERRED_ENVIRONMENT_NAME -f environment.yml
+    $ conda activate $PREFERRED_ENVIRONMENT_NAME
+
+.. note::
+
+    When using the environment file to install with Anaconda or Miniconda, the
+    name of the environment (``mne``) is built into the environment file
+    itself, but can be changed on the command line with the ``-n`` flag (as
+    shown above). This is helpful when maintaining separate environments for
+    stable and development versions of MNE-Python, or when using the
+    environment file as a starting point for new projects. See ``conda env
+    create --help`` for more info.
+
+The environment file installs the *stable* version of MNE-Python, so next we'll
+remove that and replace it with the *development* version (the clone we just
+created):
+
+.. code-block:: console
+
+    $ cd $INSTALL_LOCATION/mne-python    # make sure we're in the right folder
+    $ pip uninstall -y mne               # or: conda remove --force mne
+    $ pip install -e .                   # or: python setup.py develop
+
+.. note::
+
+    The commands ``pip install -e .`` and ``python setup.py develop`` both
+    install a python module into the current environment by creating a link to
+    the source code directory (instead of copying the code to pip's
+    ``site_packages`` directory, which is what normally happens). This means
+    that any edits you make to the MNE-Python source code will be reflected the
+    next time you open a Python interpreter and ``import mne`` (the ``-e`` flag
+    of ``pip`` stands for an "editable" installation).
```

Comment:
Un-indent, it's a critical and non-obvious point that edits take place immediately (upon interpreter restart anyway) in this mode. I'd remove mentions of `python setup.py develop`. I think `pip install -e .` is the modern way of doing it anyway.